packages/contracts/contracts/verifierHelpers/BaseVerifierHelper.sol

```diff
@@ -8,64 +8,93 @@ contract BaseVerifierHelper {
 
     uint256 public constant SNARK_SCALAR_FIELD =
         21888242871839275222246405745257275088548364400416034343698204186575808495617;
+    uint256 public immutable chainid;
 
     struct EpochKeySignals {
-        bool revealNonce;
+        uint256 epochKey;
+        uint256 stateTreeRoot;
         uint8 nonce;
         uint48 epoch;
         uint160 attesterId;
-        uint256 stateTreeRoot;
-        uint256 epochKey;
+        bool revealNonce;
+        uint48 chainId;
         uint256 data;
     }
 
     struct ReputationSignals {
-        bool proveGraffiti;
-        bool proveMinRep;
-        bool proveMaxRep;
-        bool proveZeroRep;
-        bool revealNonce;
+        uint256 epochKey;
+        uint256 stateTreeRoot;
         uint8 nonce;
         uint48 epoch;
         uint160 attesterId;
-        uint256 stateTreeRoot;
-        uint256 epochKey;
-        uint256 graffiti;
+        bool revealNonce;
+        uint48 chainId;
         uint256 minRep;
         uint256 maxRep;
+        bool proveMinRep;
+        bool proveMaxRep;
+        bool proveZeroRep;
+        bool proveGraffiti;
+        uint256 graffiti;
+        uint256 data;
     }
 
     error AttesterInvalid();
     error InvalidEpochKey();
     error InvalidProof();
     error CallerInvalid();
+    error ChainIdNotMatch(uint48 chainId);
 
     function decodeEpochKeyControl(
         uint256 control
     )
         public
         pure
         returns (
-            bool revealNonce,
-            uint160 attesterId,
+            uint8 nonce,
             uint48 epoch,
-            uint8 nonce
+            uint160 attesterId,
+            bool revealNonce,
+            uint48 chainId
         )
     {
-        uint8 attesterIdBits = 160;
-        uint8 epochBits = 48;
         uint8 nonceBits = 8;
-        revealNonce =
-            ((control >> (attesterIdBits + epochBits + nonceBits)) & 1) != 0;
-        attesterId = uint160(
-            (control >> (epochBits + nonceBits)) & ((1 << attesterIdBits) - 1)
-        );
-        epoch = uint48((control >> nonceBits) & ((1 << epochBits) - 1));
-        nonce = uint8(control & ((1 << nonceBits) - 1));
-        return (revealNonce, attesterId, epoch, nonce);
+        uint8 epochBits = 48;
+        uint8 attesterIdBits = 160;
+        uint8 revealNonceBit = 1;
+        uint8 chainIdBits = 36;
+        uint8 accBits = 0;
+
+        nonce = uint8(shift(control, accBits, nonceBits));
+        accBits += nonceBits;
+
+        epoch = uint48(shift(control, accBits, epochBits));
+        accBits += epochBits;
+
+        attesterId = uint160(shift(control, accBits, attesterIdBits));
+        accBits += attesterIdBits;
+
+        revealNonce = bool(shift(control, accBits, revealNonceBit) != 0);
+        accBits += revealNonceBit;
+
+        chainId = uint48(shift(control, accBits, chainIdBits));
+        return (nonce, epoch, attesterId, revealNonce, chainId);
+    }
+
+    function shift(
```

Comment:
Maybe call this `shiftParse` or `shiftAndParse`?

Reply:
[3639c48](https://github.com/Unirep/Unirep/pull/547/commits/3639c48267119900a32eff5b2b2b2c26d64d3241)